### PR TITLE
feat: Azure Provider HasInstance implementation

### PR DESCRIPTION
**Pull request by Bryce-Soghigian**
#### What type of PR is this?
/kind bug
/kind regression

#### What this PR does / why we need it:
CA fails to scale up or cancel in progress schaledown when there are unschedulable pods. Stealing this description from the aws provider implementation. 

I think the description of https://github.com/kubernetes/autoscaler/pull/5054#issue-1319989728 explains it well:
...original intent of determining the deleted nodes was incorrect, which led to the issues reported by other users. The nodes tainted with ToBeDeleted were misidentified as Deleted instead of Ready/Unready, which caused a miscalculation of the node being included as Upcoming. This caused problems described in https://github.com/kubernetes/autoscaler/issues/3949 and https://github.com/kubernetes/autoscaler/issues/4456.
#### Which issue(s) this PR fixes:
- https://github.com/Azure/AKS/issues/4286
- https://github.com/kubernetes/autoscaler/issues/4456

#### Special notes for your reviewer:
Lets talk a bit about what we expect HasInstance to return in different scenarios. 

The following cases we fall back to the deleted taint rather than using our has instance implementation. 

1. The providerID provided is not the provider id of an instance CAS manages. These instances will never be in the azureCache, so using it for lookups, we would see that instance as missing in the cache and return false, nil. Instead if we see an instance is unowned, we should fallback.  Note that overreporting this value here would result in us saying we are deleting many more nodes than we actually are. 
2. The provider id for the node is in an invalid structure or shape. In most cases this won't occur as other parts of the stack would also likely fail. But if we see an invalid provider id, or a standalone vm id(not vmss flex or otherwise), we fall back onto the old implementation for cloudprovider node. 

The other cases are simply true, nil, and false, nil when the instance is owned. If the instance is gone from the azure cache, we assume its been deleted and return false.  For the true case, if its in the cache we return true. Simple as that. 

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by tallaxes — COMMENTED** [review]
Overall LGTM, with minor feedback, and some comments re implication of using cache.
How was this tested? Can we add unit tests? E2E tests?

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider.go [hunk 1]
```diff
@@ -122,8 +122,12 @@ func (azure *AzureCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovid
 }
 
 // HasInstance returns whether a given node has a corresponding instance in this cloud provider
-func (azure *AzureCloudProvider) HasInstance(*apiv1.Node) (bool, error) {
-	return true, cloudprovider.ErrNotImplemented
+func (azure *AzureCloudProvider) HasInstance(node *apiv1.Node) (bool, error) {
+	ng, err := azure.azureManager.azureCache.FindForInstance(&azureRef{Name: node.Spec.ProviderID}, azure.azureManager.azureCache.vmType)
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider.go:126]: This can be reduced to
```suggestion
	ng, err := azure.azureManager.GetNodeGroupForInstance(&azureRef{Name: node.Spec.ProviderID})
```
or even further to just
```suggestion
	ng, err := azure.NodeGroupForNode(node)
```

But this brings up several questions. What exactly are the semantics of `HasInstance` and `NodeGroupForNode`? If `HasInstance` can be correctly and reliably implemented by `NodeGroupForNode` (or is it a special property of our provider?), why would `HasInstance` be needed in the CP interface to begin with? `NodeGroupForNode` returning nil is supposed to mean "instance should not be handled by autoscaler", not that it does not exist. Using it (or equivalent logic) for implementing `HasInstance` would mean NG node will always be reported as existing, non-NG node will always be reported as deleted; which seems wrong, probably need to think this through.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 2]
```diff
@@ -353,6 +359,45 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
 	return m.autoscalingOptions[ref]
 }
 
+// HasInstance returns if a given instance exists in the azure cache
+func (m *azureCache) HasInstance(instance *azureRef) (bool, error) {
+	vmsPoolSet := m.getVMsPoolSet()
```
Comment (rakechill) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:364]: Why is this only applicable to `VMsPoolSet`?

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 3]
```diff
@@ -353,6 +359,45 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
...
+	if err != nil {
+		// Most likely an invalid resource id, we should return false
+		// most of these shouldn't make it here do to higher level
+		// validation in the HasInstance azure.cloudprovider function
+		return false, err
+	}
+	inst := azureRef{Name: resourceID}
+	klog.V(4).Infof("HasInstance: Checking instance %s", inst.Name)
+
+	if m.unownedInstances[inst] {
```
Comment (rakechill) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:377]: what's the difference between `unownedInstances[]` and `isUnknownInstance()`

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 4]
```diff
@@ -353,6 +359,45 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
...
+		klog.V(4).Infof("HasInstance: Instance %s is unowned or error occurred: %v", inst.Name, err)
+		return false, cloudprovider.ErrNotImplemented
+	}
+
+	if invalid, err := m.isInvalidVMType(inst, m.vmType, vmsPoolSet); invalid || err != nil {
+		klog.V(4).Infof("HasInstance: Instance %s is invalid VM type or error occurred: %v", inst.Name, err)
+		return false, cloudprovider.ErrNotImplemented
+	}
+
+	if nodeGroup := m.getInstanceFromCache(inst.Name); nodeGroup != nil {
```
Comment (rakechill) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:392]: 1. Why are we calling the same thing in `HasInstance` + `FindForInstance`
2. Why does the log here mention `FindForInstance`

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 5]
```diff
@@ -366,42 +411,60 @@ func (m *azureCache) FindForInstance(instance *azureRef, vmType string) (cloudpr
...
+		return nil, err
+	}
+
+	if invalid, err := m.isInvalidVMType(inst, vmType, vmsPoolSet); invalid || err != nil {
+		return nil, err
+	}
+
+	// Look up caches for the instance.
+	klog.V(6).Infof("FindForInstance: attempting to retrieve instance %v from cache", m.instanceToNodeGroup)
+	if nodeGroup := m.getInstanceFromCache(inst.Name); nodeGroup != nil {
```
Comment (rakechill) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:425]: need to validate what `getInstanceFromCache` returns, since it seems to be the `nodeGroup`???

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 6]
```diff
@@ -366,42 +411,60 @@ func (m *azureCache) FindForInstance(instance *azureRef, vmType string) (cloudpr
...
+		klog.V(4).Infof("FindForInstance: found node group %q in cache", nodeGroup.Id())
+		return nodeGroup, nil
+	}
+	klog.V(4).Infof("FindForInstance: Couldn't find node group of instance %q", inst)
+	return nil, nil
+}
+
+func (m *azureCache) isUnownedInstance(inst azureRef, reason string) (bool, error) {
```
Comment (rakechill) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:433]: seems like calling this function is sufficient and HasInstance/FindForInstance do not need to also do their own `m.unownedInstances[inst]` check

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 7]
```diff
@@ -115,6 +118,10 @@ func (m *azureCache) Cleanup() {
 	close(m.interrupt)
 }
 
+func isRegisteredAKSManagedPoolName(poolName string) {
+	// Looks up if a given poolName is registered
```
Comment (rakechill) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:122]: to be implemented?

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 4 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:420]: THis is outdated, but getInstanceFromCache is just an easy way to lookup by the provider id if the instance is there.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 5 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:453]: it returns nodegroup but we don't really care what it returns just that there is an instance with a nodegroup mapping in the cache.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 6 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:461]: thats true more of a sanity check, we can probably remove that. 

unownedInstances isn't a very good way to track this anyway as it doesn't really represent all unowned instances.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 8]
```diff
@@ -137,9 +146,15 @@ func (m *azureCache) regenerate() error {
 		}
 	}
 
+	managedPoolNames := sets.New[string]()
 	// Regenerate VMSS to autoscaling options mapping.
 	newAutoscalingOptions := make(map[azureRef]map[string]string)
 	for _, vmss := range m.scaleSets {
+		// Check if the nodegroup is registered. If it is, lets store the managedPoolName
+		managedPoolName := extractAKSManagedPoolNameFromTags(vmss.Tags)
+		if registeredNodeGroupIDs.Has(managedPoolName) && managedPoolName != "" {
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:155]: need to check this logic a bit closer for registeredNodeGroupIDs and do some manual testing.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 9]
```diff
@@ -120,10 +123,12 @@ func (m *azureCache) regenerate() error {
 	if err != nil {
 		return err
 	}
-
+		
 	// Regenerate instance to node groups mapping.
 	newInstanceToNodeGroupCache := make(map[azureRef]cloudprovider.NodeGroup)
+	registeredNodeGroupIDs := sets.New[string]()
 	for _, ng := range m.registeredNodeGroups {
+		registeredNodeGroupIDs.Insert(ng.Id())
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:131]: This is either the nodepoolName or its vmss name. Need to spin up a cluster to check

Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go]: Looked on the cluster, this is the vmss.Name for vmss. Availablitysets its teh name of the availablity set. Going to ignore availability set implementation for this edge case and just proceed with vmss due to the time crunch.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 2 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:383]: I am uncertain what this question means. Can you elaborate?

File: cluster-autoscaler/cloudprovider/azure/azure_util.go [hunk 10]
```diff
@@ -647,3 +647,13 @@ func vmPowerStateFromStatuses(statuses []compute.InstanceViewStatus) string {
...
+
+func getVMSSNameFromProviderID(providerID string) string {
+	parts := strings.Split(providerID, "/")
+	for i, part := range parts {
+		if part == "virtualMachineScaleSets" && i+1 < len(parts) {
+			return parts[i+1]
+		}
+	}
+	return ""
+}
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_util.go:659]: Ideally we could move to an implementation similar to the karpenter 


SubnetIDParts https://github.com/Azure/karpenter-provider-azure/blob/main/pkg/utils/subnet_parser.go#L38. 


We should also add to that structure some ID validation rather than having providerID validation sprinkled around the codebase. We should have one standard providerID validation bit. 


Not doing in this PR but just note it bothers me

File: cluster-autoscaler/cloudprovider/azure/azure_scale_set_test.go [hunk 11]
```diff
@@ -65,6 +65,10 @@ func newTestVMSSList(cap int64, name, loc string, orchmode compute.Orchestration
 	return []compute.VirtualMachineScaleSet{
 		{
 			Name: to.StringPtr(name),
+			Tags: map[string]*string{
+				"aks-managed-poolName": to.StringPtr(name),
+			},
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_scale_set_test.go:70]: We should move this entire structure to be more high fidelity, so going to live this change in even tho we don't need it anymore after I changed to parsing scaleset from provider id

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 12]
```diff
@@ -49,33 +50,35 @@ type azureCache struct {
...
+		refreshInterval:        cacheTTL,
+		resourceGroup:          resourceGroup,
+		vmType:                 vmType,
+		vmsPoolSet:             make(map[string]struct{}),
+		scaleSets:              make(map[string]compute.VirtualMachineScaleSet),
+		virtualMachines:        make(map[string][]compute.VirtualMachine),
+		registeredNodeGroups:   make([]cloudprovider.NodeGroup, 0),
+		registeredAKSPoolNames: make(sets.Set[string]),
+		instanceToNodeGroup:    make(map[azureRef]cloudprovider.NodeGroup),
+		unownedInstances:       make(map[azureRef]bool),
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:79]: This could also be moved to a set impl, leaving it for now.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 13]
```diff
@@ -353,6 +378,46 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
 	return m.autoscalingOptions[ref]
 }
 
+// HasInstance returns if a given instance exists in the azure cache
+func (m *azureCache) HasInstance(providerID string) (bool, error) {
+	vmsPoolSet := m.getVMsPoolSet()
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	resourceID, err := convertResourceGroupNameToLower(providerID)
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:386]: this covertResourceGroupNameToLower, alongside the validation inside could be split out into our ProviderIDParser and a 
Validate() call could happen at the top of hasinstance, and here we just covert to lowercase to normalize it.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 2 above)
Comment (rakechill) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:383]: Yep, so it's unclear to me why this function is only checking VMsPool and not VMSS, as well.

Maybe I'm misunderstanding how these two nodegroup types interact? But my understanding is that `getVMsPoolSet` would fetch VMs that correspond to a VMs pool rather than a VMSS.

And I would think that this function should check for both?

File: cluster-autoscaler/cloudprovider/azure/azure_manager.go [hunk 14]
```diff
@@ -160,7 +160,6 @@ func (m *AzureManager) buildNodeGroupFromSpec(spec string) (cloudprovider.NodeGr
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse node group spec: %v", err)
 	}
-
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_manager.go:163]: why 🗿

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 6 above)
Comment (comtalyst) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:453]: > more of a sanity check

Or at least maybe there should be a comment about it if it is not too intuitive

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 15]
```diff
@@ -353,6 +378,46 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
...
+		return false, cloudprovider.ErrNotImplemented
+	}
+
+	if invalid, err := m.isInvalidVMType(inst, m.vmType, vmsPoolSet); invalid || err != nil {
+		klog.V(4).Infof("HasInstance: Instance %s is invalid VM type or error occurred: %v", inst.Name, err)
+		return false, cloudprovider.ErrNotImplemented
+	}
+
+	if nodeGroup := m.getInstanceFromCache(inst.Name); nodeGroup != nil {
+		klog.V(4).Infof("FindForInstance: found node group %q in cache", nodeGroup.Id())
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:413]: Supposed to be `"HasInstance: ..."`?

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 16]
```diff
@@ -49,33 +50,35 @@ type azureCache struct {
...
+		interrupt:              make(chan struct{}),
+		azClient:               client,
+		refreshInterval:        cacheTTL,
+		resourceGroup:          resourceGroup,
+		vmType:                 vmType,
+		vmsPoolSet:             make(map[string]struct{}),
+		scaleSets:              make(map[string]compute.VirtualMachineScaleSet),
+		virtualMachines:        make(map[string][]compute.VirtualMachine),
+		registeredNodeGroups:   make([]cloudprovider.NodeGroup, 0),
+		registeredAKSPoolNames: make(sets.Set[string]),
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:77]: Is this just to ease up the process of getting matched `strings.ToLower(registeredNodeGroup.Id())`? How much does it help, and is it really worth it?

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 17]
```diff
@@ -353,6 +378,46 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
...
+		// Most likely an invalid resource id, we should return false
+		// most of these shouldn't make it here do to higher level
+		// validation in the HasInstance azure.cloudprovider function
+		return false, err
+	}
+
+	nodeGroupNameForNode := getVMSSNameFromProviderID(providerID)
+	inst := azureRef{Name: resourceID}
+	klog.V(4).Infof("HasInstance: Checking instance %s", inst.Name)
+	if m.unownedInstances[inst] || (m.vmType == vmTypeVMSS && len(vmsPoolSet) == 0 && !m.registeredAKSPoolNames.Has(nodeGroupNameForNode)) {
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:397]: Why we don't need this in `FindForInstance()`?

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 18]
```diff
@@ -353,6 +378,46 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
...
+		klog.V(4).Infof("HasInstance: Instance %s is unowned", inst.Name)
+		return false, cloudprovider.ErrNotImplemented
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:399]: I don't think the errors should be `cloudprovider.ErrNotImplemented`, but should be real error reflecting the reason.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 19]
```diff
@@ -353,6 +378,46 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
...
+	}
+
+	if invalid, err := m.isInvalidVMType(inst, m.vmType, vmsPoolSet); invalid || err != nil {
+		klog.V(4).Infof("HasInstance: Instance %s is invalid VM type or error occurred: %v", inst.Name, err)
+		return false, cloudprovider.ErrNotImplemented
+	}
+
+	if nodeGroup := m.getInstanceFromCache(inst.Name); nodeGroup != nil {
+		klog.V(4).Infof("FindForInstance: found node group %q in cache", nodeGroup.Id())
+		return true, nil
```
Comment (comtalyst) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:414]: Why don't we use `return FindForInstance(...) != nil`?

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider.go [hunk 20]
```diff
@@ -122,8 +125,15 @@ func (azure *AzureCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovid
 }
 
 // HasInstance returns whether a given node has a corresponding instance in this cloud provider
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider.go:127]: Gave this some thought; here is what I _think_ `HasInstance` behavior should be, captured as comment. Let me know if you disagree (avoiding returning `false, nil` = "no instance" is probably the most controversial?) Need to agree on this first, as it would/should naturally be reflected in implementation and tests.

```suggestion
// HasInstance returns whether a given node has a corresponding instance in this cloud provider.
//
// Used to prevent undercount of existing VMs (taint-based overcount of deleted VMs),
// and so should not return false, nil (no instance) if uncertain; return error instead.
// (Think "has instance for sure, else error".) Returning an error causes fallback to taint-based
// determination; use ErrNotImplemented for silent fallback, any other error will be logged.
//
// Expected behavior (should work for VMSS, VMSS Flex and VMs):
// -  exists            : return true, nil
// - !exists            : return *,    NotImplemented (could use custom error for autoscaled nodes)
// - unimplemented case : return *,    NotImplemented
// - any other error    : return *,    error
```

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 18 above)
Comment (tallaxes) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:399]: Have not yet look at all the logic here, but (see my comment elsewhere) `ErrNotImplemented` causes _silent_ fallback, which would likely be appropriate for unowned instances. Otherwise, in presence of non-autoscaled nodes, this would be spamming the logs all the time.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 21]
```diff
@@ -170,26 +177,44 @@ func (m *azureCache) regenerate() error {
...
+	for key, val := range vmssTags {
+		if val == nil {
+			continue
+		}
+		if key == agentpoolNameTag {
+			return *val
+		}
+	}
+	return ""
+}
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:190]: This can be removed now too

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 16 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:77]: providerID normalization seems very worth it in this case. Some apis return it in different casing so lowercasing it seems logical to me.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 17 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:397]: We potentially could need this for FindForInstance. I am just limiting the blast radius of this change.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 18 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:399]: @comtalyst  see silent fallback here https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/clusterstate/clusterstate.go#L1055

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 15 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:413]: Good catch

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 19 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:414]: There is some different logic for checking for unowned instances.

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 22]
```diff
@@ -187,77 +313,66 @@ func TestMixedNodeGroups(t *testing.T) {
...
-			},
-		}
-		group, err = provider.NodeGroupForNode(nodeNotInGroup)
-		assert.NoError(t, err)
-		assert.Nil(t, group)
-	}
-
-}
-
-func TestNodeGroupForNodeWithNoProviderId(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go:247]: Accident, added it back.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 2 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:371]: This implementation falls back to the old IsDeleting logic for any non vmss pools due to urgency.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 23]
```diff
@@ -140,6 +140,7 @@ func (m *azureCache) regenerate() error {
 	// Regenerate VMSS to autoscaling options mapping.
 	newAutoscalingOptions := make(map[azureRef]map[string]string)
 	for _, vmss := range m.scaleSets {
+		// Check if the nodegroup is registered. If it is, lets store the managedPoolName
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:143]:
```suggestion
```

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 24]
```diff
@@ -366,42 +392,60 @@ func (m *azureCache) FindForInstance(instance *azureRef, vmType string) (cloudpr
...
 	}
 	inst := azureRef{Name: resourceID}
+
+	if unowned, err := m.isUnownedInstance(inst, ""); unowned || err != nil {
+		return nil, err
+	}
+
+	if invalid, err := m.isInvalidVMType(inst, vmType, vmsPoolSet); invalid || err != nil {
+		return nil, err
+	}
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:402]: Please either undo this refactoring, or replace it with separate helpers for checking vs. marking as unowned:
* Neither `isUnownedInstance` nor `isInvalidVMType` return errors, no reason for (bool,error) signature
* It is very unintuitive/unexpected for `is*` functions to have a side effect (marking instance as unowned in this case); making it conditional on special value of second parameter in `isUnownedInstance` makes it worse

File: cluster-autoscaler/cloudprovider/azure/azure_scale_set_test.go (hunk 11 above)
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_scale_set_test.go:70]: Is this change still relevant/needed?

File: cluster-autoscaler/cloudprovider/azure/azure_util.go [hunk 25]
```diff
@@ -647,3 +647,13 @@ func vmPowerStateFromStatuses(statuses []compute.InstanceViewStatus) string {
 	// PowerState is not set if the VM is still creating (or has failed creation)
 	return vmPowerStateUnknown
 }
+
+func getVMSSNameFromProviderID(providerID string) string {
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_util.go:651]: func `getVMSSNameFromProviderID` is unused

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 26]
```diff
@@ -187,61 +313,68 @@ func TestMixedNodeGroups(t *testing.T) {
...
+			assert.Equal(t, group.MinSize(), 1)
+			assert.Equal(t, group.MaxSize(), 5)
+
+			// test node in cluster that is not in a group managed by cluster autoscaler
+			nodeNotInGroup := &apiv1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						azureAgentpoolKey: "unregistered-friend",
+					},
+				},
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go:363]: Why is this needed?

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 27]
```diff
@@ -353,6 +359,26 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
 	return m.autoscalingOptions[ref]
 }
 
+// HasInstance returns if a given instance exists in the azure cache
+func (m *azureCache) HasInstance(providerID string) (bool, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	resourceID, err := convertResourceGroupNameToLower(providerID)
+	if err != nil {
+		// Most likely an invalid resource id, we should return false
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:368]: "should return error"? Actual return value does not matter in the presence of error

**Review by tallaxes — CHANGES_REQUESTED** [review]
The actual implementation looks good to me; requested changes are elsewhere.

Please also update the expected `HasInstance` behavior in the description.

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go [hunk 28]
```diff
@@ -187,61 +313,68 @@ func TestMixedNodeGroups(t *testing.T) {
...
+			}
+			group, err = provider.NodeGroupForNode(nodeNotInGroup)
+			assert.NoError(t, err)
+			assert.Nil(t, group)
+
+			hasInstance, err := provider.HasInstance(nodeNotInGroup)
+			assert.False(t, hasInstance)
+			assert.Error(t, err)
+			assert.Equal(t, err, cloudprovider.ErrNotImplemented)
```
Comment (tallaxes) [inline cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go:375]: Maybe add a corresponding positive test above as well? (Though mixing testing `NodeGroupForNode` and `HasInstance` is not great ...)

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go (hunk 26 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go:363]: It is a remnant from the previous implementations. We can remove it.

File: cluster-autoscaler/cloudprovider/azure/azure_scale_set_test.go (hunk 11 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_scale_set_test.go:70]: no.

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 24 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:401]: Removed

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go (hunk 27 above)
Comment (Bryce-Soghigian) [inline cluster-autoscaler/cloudprovider/azure/azure_cache.go:368]:
```suggestion
		// Most likely an invalid resource id, we should return an error
```

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider.go [hunk 29]
```diff
@@ -33,6 +34,8 @@ import (
 const (
 	// GPULabel is the label added to nodes with GPU resource.
 	GPULabel = "accelerator"
+
+	azureAgentpoolKey = "kubernetes.azure.com/agentpool"
```
[the remark this answers is not in the document]
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cloud_provider.go]: removed this

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go (hunk 28 above)
Comment (Bryce-Soghigian) [reply cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go]: Added a positive one. Should we factor out this into two T.Run() cases one for HasInstance and another for NodeGroupForNode where they both share the setup?

File: cluster-autoscaler/cloudprovider/azure/azure_cache.go [hunk 30]
```diff
@@ -353,6 +359,26 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
...
+		// most of these shouldn't make it here do to higher level
+		// validation in the HasInstance azure.cloudprovider function
+		return false, err
+	}
+
+	inst := azureRef{Name: resourceID}
+	if nodeGroup := m.getInstanceFromCache(inst.Name); nodeGroup != nil {
+		return true, nil
+	}
+	// couldn't find instance in the cache, assume its deleted
```
[the remark this answers is not in the document]
Comment (tallaxes) [reply cluster-autoscaler/cloudprovider/azure/azure_cache.go:377]: Not applied?

File: cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go (hunk 28 above)
Comment (tallaxes) [reply cluster-autoscaler/cloudprovider/azure/azure_cloud_provider_test.go]: No, much of the setup would still have to be replicated? Probably not critical, given other tests for HasInstance.
